### ISSUE TITLE
Custom event type facades

### ISSUE THREAD

**Issue by buntec** [raquo/laminar-shoelace-components#2]
First of all, I wanted to say thank you for this library (and domtypes, of course, which made writing my own UI lib possible in the first place)! 

I've been trying to [adapt it](https://github.com/buntec/ff4s-shoelace) to a UI lib I've been working on. Everything works great so far except for one little thing: I think the facades for custom event types don't quite work b/c custom fields should be nested inside a `detail` field. See [here](https://developer.mozilla.org/en-US/docs/Web/Events/Creating_and_triggering_events#adding_custom_data_%E2%80%93_customevent). There is a nice solution for this [here](https://github.com/sherpal/LaminarSAPUI5Bindings/blob/c56abfd7ddfb2116675bc1f0dbe44680c260171f/web-components/src/main/scala/be/doeraene/webcomponents/ui5/Tree.scala#L50).  I've managed to work around it by changing the source generator ever so slightly. (I also really like the `EventWithPreciseTarget` idea from that project, which I've also adopted).

In the long run, it would be really great to see the codegen factored out into a lib. I'd be happy to help!

**Comment by raquo** [comment]
Yeah that would be nice to have. I didn't realize the `detail` field was an actual standard, I thought it was just a UI5 thing. But looks like even Shoelace uses it for some events...

If you can point me to the changes you made to support this use case, maybe I'll be able to incorporate them eventually, but just to set the expectations – no promises, especially in terms of timeline. My focus is just elsewhere right now, I don't know when I'll be able to get back to Shoelace.

**Comment by buntec** [comment]
All relevant changes are in [this commit](https://github.com/buntec/ff4s-shoelace/commit/2194d300a8ff98ffdbfd47527e0b2725cf70a626). The naming of the custom event types should probably be changed to something like `CopyEventDetail`, `ErrorEventDetail`, etc. to reflect the new usage.

**Comment by LaurentLegrand** [comment]
Hello!

Here is a patch of ShoalaceGenerator.scala that generates a `Detail` trait for each event ; trait located in the object companion. For instance:

```scala
  @js.native
  trait SelectionChangeEvent extends CustomEvent {

    val detail: SelectionChangeEvent.Detail

  }
  object SelectionChangeEvent {

    @js.native
    trait Detail extends js.Object {

      val selection: js.Array[TreeItem.Ref]
    }
  }

```
It works at least for Tree and TreeItem.

```
diff --git a/project/ShoelaceGenerator.scala b/project/ShoelaceGenerator.scala
index f4cf041..feba299 100644
--- a/project/ShoelaceGenerator.scala
+++ b/project/ShoelaceGenerator.scala
@@ -183,11 +183,21 @@ class ShoelaceGenerator(
         line()
         line("@js.native")
         enter(s"trait ${et.scalaName} extends ${baseCustomEventType} {", "}") {
-          et.fields.foreach { field =>
-            val scalaTypeStr = st.scalaPropOutputType(s"prop `${field.domName}` in event type `${et.scalaName}`", field.jsTypes)
-            line()
-            blockCommentLines(field.description)
-            line(s"val ${field.scalaName}: ${scalaTypeStr}")
+          line()
+          line(s"val detail: ${et.scalaName}.Detail")
+          line()
+        }
+
+        enter(s"object ${et.scalaName} {", "}") {
+          line()
+          line("@js.native")
+          enter(s"trait Detail extends js.Object {", "}") {
+            et.fields.foreach { field =>
+              val scalaTypeStr = st.scalaPropOutputType(s"prop `${field.domName}` in event type `${et.scalaName}`", field.jsTypes)
+              line()
+              blockCommentLines(field.description)
+              line(s"val ${field.scalaName}: ${scalaTypeStr}")
+            }
           }
         }
         et.scalaName

```
